securedrop/journalist_app/api2/__init__.py

```diff
@@ -0,0 +1,148 @@
+import hashlib
+from typing import Any, Dict, Optional
+
+from flask import Blueprint, abort, json, jsonify, request
+from models import Source
+from sqlalchemy.inspection import inspect
+from sqlalchemy.orm import Query, joinedload
+from werkzeug.wrappers.response import Response
+
+blp = Blueprint("api2", __name__, url_prefix="/api/v2")
+
+PREFIX_MAX_LEN = inspect(Source).columns["uuid"].type.length
+
+
+def all_sources() -> Query:
+    """
+    Return a base query for all ``Sources`` with eager loading of their metadata
+    and collections.
+    """
+    return (
+        Source.query.options(joinedload(Source.star))
+        .options(joinedload(Source.submissions))
+        .options(joinedload(Source.replies))
+    )
+
+
+def json_version(d: dict) -> str:
+    """
+    Calculate the version (BLAKE2s digest) of the normalized JSON representation
+    of the dictionary ``d``.
+
+    We use BLAKE2s here because SHA-256 is too slow (we don't care about
+    cryptographic security) and CRC-32 is too collision-prone (we're not merely
+    checksumming for transmission integrity).
+    """
+    s = json.dumps(d, sort_keys=True)
```

Comment:
can we add `separators=[',', ':']` to the call here?

it turns out that python `json.dumps` and js `JSON.stringify` create slightly different outputs. namely, `json.dumps` does some pretty-printing with whitespace so in order to actually get matching versions, the simplest way (i think) is to remove the whitespace from the python json stringification